### ISSUE TITLE
Updating a poco record, with specific primaryid

### ISSUE THREAD

**Issue by benosman** [schotime/NPoco#285]
I'm trying to update a poco record from a mvc controller edit form, but i'm having trouble using the function with the signature: 
```c# 
public int Update(object poco, object primaryKeyValue)
```

I get the sql error: "Cannot update identity column 'id' "

The model is returned from the form submission with the primary key set to 0 (as the field is not present in the form).

So, the function tries to override the primary key value, but that still returns the above error.

```c#        
var newId = Database.Update(poco, id);
```

I traced through the code and it determines the correct primary key name, but the logic in the following code doesn't seem to account for the situation where the primary key value is not null.

https://github.com/schotime/NPoco/blob/1b40e272e22269ef2c29bde70d3093dc97ab4cc2/src/NPoco/Database.cs#L1502-L1506

To me the code should be more like this, unless i'm missing something obvious here.
```c#                
if (primaryKeyValuePairs.ContainsKey(pocoColumn.ColumnName))
    if (primaryKeyValue == null)
         primaryKeyValuePairs[pocoColumn.ColumnName] = ProcessMapper(pocoColumn, pocoColumn.GetValue(poco));
    continue;
}
```

**Comment by schotime** [comment]
How can you do an update of a row with a null primary key value?

**Comment by benosman** [comment]
Obviously I didn't explain the issue well enough, but as you know that is not possible.

The primary key on the model itself is the default value of  0 (from a form submission).

Since at this point the model could be one of several types, I don't automatically know the primary key name.

So I want to use Database.Update to set a specific id value (passed in as a parameter and certainly not null). 

Thats where it falls down and I get the error.

**Comment by schotime** [comment]
This still doesn't make any sense sorry.

**Comment by benosman** [comment]
One last try - can you tell me what the purpose of[ this method overload](
https://github.com/schotime/NPoco/blob/1b40e272e22269ef2c29bde70d3093dc97ab4cc2/src/NPoco/Database.cs#L1661) is:

```c# 
        public int Update(object poco, object primaryKeyValue)
        {
            return Update(poco, primaryKeyValue, null);
        }
```

I've been assuming its a way to save a poco object and override the primaryKeyValue with the one in the parameter.

If my assumption is correct - it does not work as expected, which is the reason for my post.

**Comment by schotime** [comment]
This method is to update the Poco passed to it where the primarykey value is the value passed in.

eg.

`db.Update(new User() { Name = "New Name" }, 2)`
will execute something like
`update users set name = @0 where id = @1`
where `@0` comes from the `User` object (New Name) and the `@1` comes from the value passed in (2).

**Comment by benosman** [comment]
Ok, that makes sense.

The problem I have is that i'm trying to do that, but on the poco model in question the primary key property is an int value (e.g. non null). 

So creating the new instance, without explicitly setting the primary key value, sets the primary key value to 0. So for your example the following is executed:
```
UPDATE users SET id = @0, name = @1 WHERE id = @2
```
@0 = 0, @1 = "New Name", @2 = 2

That is where I get the error  "Cannot update identity column 'id' ", which makes sense of course given the SQL.

What I am proposing, is when updating using a specific primary key value, that any primary key values present on the instance are excluded from the set part of the update query.

**Comment by schotime** [comment]
Ok, what that probably means is that id isn't configured as your primary key.
Can you show me what your class looks like and how the primary key is configured?

**Comment by benosman** [comment]
It is set correctly as the primary key.

```c#
    [PrimaryKey("contactid")]
    [ExplicitColumns]
    public partial class Contact
    {
		[Column]
		public int contactid 
		{ 
			get
			{
				return _contactid;
			}
			set
			{
				_contactid = value;
			}
		}
		int _contactid;
    }
```

What's more, when stepping though the code in the debugger, I can see it locates the primary key column correctly (the set part is correct, and I didn't specify the column name).

**Comment by benosman** [comment]
the logic in [Database.cs](https://github.com/schotime/NPoco/blob/1b40e272e22269ef2c29bde70d3093dc97ab4cc2/src/NPoco/Database.cs#L1502-L1506) currently goes as follows:

```c#

// Don't update the primary key, but grab the value if we don't have it
if (primaryKeyValue == null && primaryKeyValuePairs.ContainsKey(pocoColumn.ColumnName))
{
    primaryKeyValuePairs[pocoColumn.ColumnName] = ProcessMapper(pocoColumn, pocoColumn.GetValue(poco));
    continue;
}

<snip> Other unrelated checks and the actual adding of the column to the sql and arguments</snip>

```
In the code above, it only skips adding the value to the update column list IF primary key value **IS NULL** AND the primary Key column is found in primaryKeyValuePairs.

In my case the primary key value is not null and so it just adds the column to the Update SQL.

As shown in the first post, i propose first checking if the primary column is found in primaryKeyValuePairs and if so skip adding to the Update sql. Then inside that condition the check if the primaryKeyValue is null, if it is then the value is retrieved from the instance itself.

```c# 
if (primaryKeyValuePairs.ContainsKey(pocoColumn.ColumnName))
    if (primaryKeyValue == null)
         primaryKeyValuePairs[pocoColumn.ColumnName] = ProcessMapper(pocoColumn, pocoColumn.GetValue(poco));
    continue;
}
```

**Comment by schotime** [comment]
Makes sense. Want to make a PR with a test?

**Comment by schotime** [comment]
This test will reproduce it.

```
[Test]
        public void UpdatePrimaryKeyObjectOverridingPrimaryKey()
        {
            var poco = Database.SingleOrDefaultById<UserDecorated>(InMemoryUsers[1].UserId);
            Assert.IsNotNull(poco);

            poco.Age = InMemoryUsers[1].Age + 100;
            poco.Savings = (Decimal)1234.23;
            Database.Update(poco, InMemoryUsers[2].UserId);

            var verify = Database.SingleOrDefaultById<UserDecorated>(InMemoryUsers[2].UserId);
            Assert.IsNotNull(verify);

            Assert.AreEqual(InMemoryUsers[2].UserId, verify.UserId);
            Assert.AreNotEqual(InMemoryUsers[2].Age, verify.Age);
            Assert.AreNotEqual(InMemoryUsers[2].Savings, verify.Savings);
        }
```

It can go in the UpdateTests under `UpdatePrimaryKeyObject()`

**Comment by schotime** [comment]
I have fixed this and will be deployed in 3.4.0
